src/DynamoCore/Graph/Workspaces/SerializationConverters.cs

```diff
@@ -63,18 +63,30 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
             var outPorts = obj["Outputs"].ToArray().Select(t => t.ToObject<PortModel>()).ToArray();
 
             var resolver = (IdReferenceResolver)serializer.ReferenceResolver;
+            string assemblyLocation = objectType.Assembly.Location;
 
-            if (type == typeof(Function))
+            bool remapPorts = true;
+            if (type == null)
+            {
+                node = CreateDummyNode(obj, assemblyLocation, inPorts, outPorts);
+            }
+            else if (type == typeof(Function))
             {
                 var functionId = Guid.Parse(obj["FunctionUuid"].Value<string>());
-                node = manager.CreateCustomNodeInstance(functionId);
-                RemapPorts(node, inPorts, outPorts, resolver);
+
+                CustomNodeDefinition def = null;
+                CustomNodeInfo info = null;
+                bool isUnresolved = !manager.TryGetCustomNodeData(functionId, null, false, out def, out info);
+                Function function = manager.CreateCustomNodeInstance(functionId, null, false, def, info);
+                node = function;
+
+                if (isUnresolved)
+                  function.UpdatePortsForUnresolved(inPorts, outPorts);
             }
             else if (type == typeof(CodeBlockNodeModel))
             {
                 var code = obj["Code"].Value<string>();
                 node = new CodeBlockNodeModel(code, guid, 0.0, 0.0, libraryServices, ElementResolver);
-                RemapPorts(node, inPorts, outPorts, resolver);
```

Comment:
this is not required because of nodemodel changes?

Reply:
The only time that calling RemapPorts is not required is when there is a JSON constructor that forwards the ports. There were many calls to RemapPorts that have been consolidated to after the big if/elseif/.../else statement to simplify the code. (In other words the call to RemapPorts for CodeBlockNodeModel nodes is still being made, just in a different place. See line 134.)